lib/Monotone_framework.ml

```diff
@@ -0,0 +1,942 @@
+            Map.add_exn accum ~key
+              ~data:
+                { entry= analysis_in_data
+                ; exit= T.transfer_function key analysis_in_data } )
+          F.successors
+      in
+      analysis_in_out
+  end
+  : MONOTONE_FRAMEWORK
+    with type labels = l and type properties = p )
+
+let rec declared_variables_stmt
+    (s : (Mir.expr_typed_located, Mir.stmt_loc) Mir.statement) =
+  match s with
+  | Mir.Decl {decl_id= x; _} -> Set.Poly.singleton x
+  | Mir.Assignment (_, _)
+   |Mir.TargetPE _
+   |Mir.NRFunApp (_, _)
+   |Mir.Check _ | Mir.Break | Mir.Continue | Mir.Return _ | Mir.Skip ->
+      Set.Poly.empty
+  | Mir.IfElse (_, b1, Some b2) ->
+      Set.Poly.union
+        (declared_variables_stmt b1.stmt)
+        (declared_variables_stmt b2.stmt)
+  | Mir.While (_, b) | Mir.IfElse (_, b, None) ->
+      declared_variables_stmt b.stmt
+  | Mir.For {loopvar= s; body= b; _} ->
+      Set.Poly.add (declared_variables_stmt b.stmt) s
+  | Mir.Block l | Mir.SList l ->
+      Set.Poly.union_list
+        (List.map ~f:(fun x -> declared_variables_stmt x.stmt) l)
+  | Mir.FunDef {fdname= f; fdargs= l; fdbody= b; _} ->
+      Set.Poly.union
+        (Set.Poly.of_list (f :: List.map l ~f:(fun (_, x, _) -> x)))
+        (declared_variables_stmt b.stmt)
+
+let propagation_mfp (prog : Mir.typed_prog)
+    (module Flowgraph : Monotone_framework_sigs.FLOWGRAPH
+      with type labels = int)
+    (flowgraph_to_mir : (int, Mir.stmt_loc_num) Map.Poly.t)
+    (propagation_transfer :
+         (int, Mir.stmt_loc_num) Map.Poly.t
+      -> (module
+          TRANSFER_FUNCTION
+            with type labels = int
+             and type properties = (string, Mir.expr_typed_located) Map.Poly.t
+                                   option)) =
+  let mir = Map.find_exn flowgraph_to_mir 1 in
+  let domain =
+    ( module struct
+      type vals = string
+
+      let total =
+        Set.Poly.union_list
+          [ Set.Poly.of_map_keys prog.gen_quant_vars
+          ; Set.Poly.of_map_keys prog.tdata_vars
+          ; Set.Poly.of_map_keys prog.tparams
+          ; Set.Poly.of_map_keys prog.params
+          ; Set.Poly.of_map_keys prog.data_vars
+          ; declared_variables_stmt
+              (Mir.stmt_loc_of_stmt_loc_num flowgraph_to_mir mir).stmt ]
+    end
+    : TOTALTYPE
+      with type vals = string )
+  in
+  let codomain =
+    (module struct type vals = Mir.expr_typed_located
+    end
+    : TYPE
+      with type vals = Mir.expr_typed_located )
+  in
+  let (module Lattice) =
+    dual_partial_function_lattice_with_bot domain codomain
+  in
+  let (module Transfer) = propagation_transfer flowgraph_to_mir in
+  let (module Mf) =
+    monotone_framework (module Flowgraph) (module Lattice) (module Transfer)
+  in
+  Mf.mfp ()
+
+let reaching_definitions_mfp (mir : Mir.typed_prog)
+    (module Flowgraph : Monotone_framework_sigs.FLOWGRAPH
+      with type labels = int)
+    (flowgraph_to_mir : (int, Mir.stmt_loc_num) Map.Poly.t) =
```

Comment:
I have a could of questions/concerns about this signature and some similar ones:
- `flowgraph_to_mir` is has lots of redundancy, since each label stores the entire corresponding subtree, right? There's another type, `(label, ('e, label) statement * 'm) Map.Poly.t` that's already constructed when Flowgraph is constructed, and doesn't have redundancy, and can also be converted back into the normal mir via `build_recursive_statement`
- It seems like `Flowgraph` and `flowgraph_to_mir` can both be directly computed from `mir`, and are therefore redundant information, as far as I can tell. That makes me wonder if this could instead just take the `mir` and compute `Flowgraph` and `flowgraph_to_mir`, or if it could take `Flowgraph`, `flowgraph_to_mir`, and `initial`

Reply:
On the first point, `flowgraph_to_mir : (int, Mir.stmt_loc_num) Map.Poly.t`, not `flowgraph_to_mir : (int, Mir.stmt_loc) Map.Poly.t`, so it shouldn't be storing subtrees, right?

On your second point, definitely! That is a refactor we should do.

Reply:
Oh, `Mir.stmt_loc_num` isn't recursive on statements, got it. I lose track of the names of the various combinations of expression and statement types, but they seem necessary under the current scheme. I think it would be really nice to have a type where the type variables actually corresponded to the metadata, but that change might not be in the cards.

Reply:
I wouldn't exclude the possibility of that change. It's mostly that everything would need to be refactored then, which is some work.